mhkit/acoustics/base.py

```diff
@@ -0,0 +1,448 @@
+"""
+This file includes the main passive acoustics analysis functions. They
+are designed to function on top of one another, starting from reading
+in wav files from the io submodule.
+"""
+
+import warnings
+import numpy as np
+import xarray as xr
+
+from mhkit.dolfyn import VelBinner
+from mhkit.dolfyn.time import epoch2dt64, dt642epoch
+
+
+def _fmax_warning(fn, fmax):
+    """
+    Check that max frequency limit isn't greater than Nyquist frequency
+    """
+
+    if fmax > fn:
+        warnings.warn(
+            "`fmax` = {fmax} is greater than the Nyquist frequency. Setting"
+            "fmax = {fn}"
+        )
+        fmax = fn
+
+    return fmax
+
+
+def minimum_frequency(water_depth, c=1500, c_seabed=1700):
+    """
+    Estimate the shallow water cutoff frequency based on the speed of
+    sound in the water column and the speed of sound in the seabed
+    material (generally ranges from 1450 - 1800 m/s)
+
+    Jennings 2011 - Computational Ocean Acoustics, 2nd ed.
+    """
+
+    f_min = c / (4 * water_depth * np.sqrt(1 - (c / c_seabed) ** 2))
+
+    return f_min
+
+
+def sound_pressure_spectral_density(pressure, fs, window=1):
+    """
+    Calculates the mean square sound pressure spectral density from audio
+    samples split into FFTs with a specified window_size in seconds and
+    at least a 50% overlap. The amplitude of the PSD is adjusted
+    according to Parseval's theorem.
+
+    Parameters
+    ----------
+    pressure: xarray.DataArray (time)
+        Sound pressure in [Pa] or volts [V]
+    fs: int
+        Data collection sampling rate [Hz]
+    window: string (optional)
+        Length of time in seconds to create FFTs. Default: 1 s.
+
+    Returns
+    -------
+    spsd: xarray.DataArray (time, freq)
+        Spectral density [Pa^2/Hz] indexed by time and frequency
+    """
+
+    # window length of each time series
+    win = window * fs
+
+    # Use dolfyn PSD
+    binner = VelBinner(n_bin=win, fs=fs, n_fft=win)
+    # Always 50% overlap if numbers reshape perfectly
```

Comment:
Yes, you'd have 19 segments, from which you're calculating periodograms for each. Segments and windows are different concepts.
I'm not an expert in fast fourier transforms, but here is the original document for Welch's method for power spectra, which is the mathematical explanation for that dolfyn function: https://ieeexplore.ieee.org/abstract/document/1161901

Reply:
Okay if it should be 19 then I don't its not working correctly.  Can you review this test?

```
def test_sound_pressure_spectral_density(self):
    """
    Test sound pressure spectral density calculation.
    """
    # Create some sample pressure data
    time = np.arange(0, 10, 0.1)
    data = np.sin(time)
    pressure = xr.DataArray(data, coords=[time], dims=["time"], attrs={"units": "Pa", "fs": 100})

    # Adjust window size to get multiple segments
    fs = 100
    window = 0.1  # seconds
    win_samples = int(window * fs)

    # Run the spectral density function
    spsd = acoustics.sound_pressure_spectral_density(pressure, fs=fs, window=window)

    # Assert that output is an xarray DataArray with expected dimensions
    self.assertIsInstance(spsd, xr.DataArray)
    self.assertIn("freq", spsd.dims)
    self.assertIn("time", spsd.dims)
    self.assertEqual(spsd.attrs["units"], "Pa^2/Hz")
    self.assertEqual(spsd.attrs["window"], f"{window}s")

    # Calculate expected number of segments
    overlap = 0.5
    step = int(win_samples * (1 - overlap))
    expected_segments = (len(pressure) - win_samples) // step + 1

    self.assertEqual(spsd.shape[0], expected_segments)
```

Reply:
No, sorry, I mispoke it seems. It's been a long time since I've dug into it, but windows are basically smoothing functions (or filters) of a specified length that are applied atop of the signal. The 50% part comes in based on the amount of data the window, i.e. smoothing function, covers.

I'm digging into textbooks at this point to try and explain it to you.

Reply:
I replaced all the instances of "window" with "bin" and updated the docstring to try and avoid confusion